From: Jiebin Sun <jiebin.sun@intel.com>
Subject: ipc/msg: avoid negative value by overflow in msginfo
Date: Tue, 20 Sep 2022 23:08:09 +0800

The 32-bit value in msginfo struct could be negative if we get it from
signed 64-bit.  Clamping it to INT_MAX helps to avoid the negative value
by overflow.

Link: https://lkml.kernel.org/r/20220920150809.4014944-1-jiebin.sun@intel.com
Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>
Reviewed-by: Manfred Spraul <manfred@colorfullif.com>
Reviewed-by: Tim Chen <tim.c.chen@linux.intel.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---


--- a/ipc/msg.c~ipc-msg-mitigate-the-lock-contention-with-percpu-counter-fix
+++ a/ipc/msg.c
@@ -501,8 +501,8 @@ static int msgctl_info(struct ipc_namesp
 	max_idx = ipc_get_maxidx(&msg_ids(ns));
 	up_read(&msg_ids(ns).rwsem);
 	if (cmd == MSG_INFO) {
-		msginfo->msgmap = percpu_counter_sum(&ns->percpu_msg_hdrs);
-		msginfo->msgtql = percpu_counter_sum(&ns->percpu_msg_bytes);
+		msginfo->msgmap = min(percpu_counter_sum(&ns->percpu_msg_hdrs), INT_MAX);
+		msginfo->msgtql = min(percpu_counter_sum(&ns->percpu_msg_bytes), INT_MAX);
 	} else {
 		msginfo->msgmap = MSGMAP;
 		msginfo->msgpool = MSGPOOL;
_
